internal/pubsub/awspubsub/awspubsub.go

```diff
@@ -0,0 +1,202 @@
+		for k, v := range dm.Metadata {
+			attrs[k] = &sns.MessageAttributeValue{
+				DataType:    aws.String("String"),
+				StringValue: aws.String(v),
+			}
+		}
+		params := sns.PublishInput{
+			Message:           aws.String(string(dm.Body)),
+			MessageAttributes: attrs,
+			TopicArn:          &t.arn,
+		}
+		if err := t.send(&params); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *topic) send(params *sns.PublishInput) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	req, _ := t.client.PublishRequest(params)
+	return req.Send()
+}
+
+// IsRetryable implements driver.Topic.IsRetryable.
+func (t *topic) IsRetryable(error) bool {
+	// The client handles retries.
+	return false
+}
+
+// As implements driver.Topic.As.
+func (t *topic) As(i interface{}) bool {
+	c, ok := i.(**sns.SNS)
+	if !ok {
+		return false
+	}
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	*c = t.client
+	return true
+}
+
+type subscription struct {
+	mu     sync.Mutex
+	client *sqs.SQS
+	qURL   string
+}
+
+// SubscriptionOptions will contain configuration for subscriptions.
+type SubscriptionOptions struct{}
+
+func OpenSubscription(ctx context.Context, client *sqs.SQS, qURL string, opts *SubscriptionOptions) *pubsub.Subscription {
+	ds := openSubscription(ctx, client, qURL)
+	return pubsub.NewSubscription(ds)
+}
+
+// openSubscription returns a driver.Subscription.
+func openSubscription(ctx context.Context, client *sqs.SQS, qURL string) driver.Subscription {
+	return &subscription{client: client, qURL: qURL}
+}
+
+// ReceiveBatch implements driver.Subscription.ReceiveBatch.
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	output, err := s.receiveMessages()
+	if err != nil {
+		return nil, err
+	}
+	var ms []*driver.Message
+	for _, m := range output.Messages {
+		type Attribute struct {
+			Value string
+		}
+		type MsgBody struct {
+			Message           string
+			MessageAttributes map[string]Attribute
+		}
+		var body MsgBody
+		if err := json.Unmarshal([]byte(*m.Body), &body); err != nil {
+			return nil, err
+		}
+		attrs := map[string]string{}
+		for k, v := range body.MessageAttributes {
+			attrs[k] = v.Value
+		}
+		m := &driver.Message{
+			Body:     []byte(body.Message),
+			Metadata: attrs,
+			AckID:    *m.ReceiptHandle,
+		}
+		ms = append(ms, m)
+	}
+	return ms, nil
+}
+
+func (s *subscription) receiveMessages() (*sqs.ReceiveMessageOutput, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	params := sqs.ReceiveMessageInput{QueueUrl: &s.qURL}
+	req, output := s.client.ReceiveMessageRequest(&params)
+	err := req.Send()
+	return output, err
+}
+
+// SendAcks implements driver.Subscription.SendAcks.
+func (s *subscription) SendAcks(ctx context.Context, ids []driver.AckID) error {
+	for _, id := range ids {
+		rh := id.(string)
+		err := s.deleteMessage(rh)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *subscription) deleteMessage(receiptHandle string) error {
```

Comment:
After you get rid of the lock, you can just inline this into the above function.

Reply:
I get a race when running the test in replace mode if I remove the mutex for the subscription client, probably because fields within the client are being mutated.

Reply:
This is surprising considering that the docs say SQS clients are safe for concurrent use: https://docs.aws.amazon.com/sdk-for-go/api/service/sqs/.

Reply:
What's the actual data that's racy?